### PR TITLE
Add Notification class, that holds info for concrete notification

### PR DESCRIPTION
**Pull request by repitch**
Now it's possible to change textColor and backgroundColor for each notification individually
Notification = text + textColor + backgroundColor

### CODE REVIEW

File: ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java [hunk 1]
```diff
@@ -152,7 +155,7 @@ protected Parcelable onSaveInstanceState() {
 		Bundle bundle = new Bundle();
 		bundle.putParcelable("superState", super.onSaveInstanceState());
 		bundle.putInt("current_item", currentItem);
-		bundle.putStringArray("notifications", notifications);
+        bundle.putParcelableArrayList("notifications", (ArrayList<Notification>) notifications);
```
Comment (seva91) [inline ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java]: It's unsafe to cast your List to ArrayList. 
If you change Notification.generateEmptyList(..) realization it's possible to get ClassCastException
Use new ArrayList(notifications);

File: ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java [hunk 2]
```diff
@@ -1257,19 +1264,37 @@ public void setNotification(int nbNotification, int itemPosition) {
...
+	 * Set Notification text
 	 *
 	 * @param title        String
 	 * @param itemPosition int
 	 */
 	public void setNotification(String title, int itemPosition) {
-		notifications[itemPosition] = title;
-		updateNotifications(false, itemPosition);
+        if (itemPosition < 0 || itemPosition > items.size() - 1) {
+            throw new RuntimeException("The position is out of bounds of the items (" + items.size() + " elements)");
```
Comment (seva91) [inline ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java]: it's better to use IllegalArgumentException in this case

File: ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java [hunk 3]
```diff
@@ -1257,19 +1264,37 @@ public void setNotification(int nbNotification, int itemPosition) {
...
+
+    /**
+     * Set fully customized Notification
+     *
+     * @param notification Notification
+     * @param itemPosition int
+     */
+	public void setNotification(Notification notification, int itemPosition) {
+		if (itemPosition < 0 || itemPosition > items.size() - 1) {
+			throw new RuntimeException("The position is out of bounds of the items (" + items.size() + " elements)");
```
Comment (seva91) [inline ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java]: it's better to use IllegalArgumentException in this case

File: ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java (hunk 1 above)
Comment (repitch) [inline ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java]: fixed in 10468b7

File: ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java (hunk 2 above)
Comment (repitch) [inline ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java]: fixed in 10468b7

File: ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java (hunk 3 above)
Comment (repitch) [inline ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java]: fixed in 10468b7

**Comment by repitch** [comment]
@seva91 just fixed your comments

File: ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java [hunk 4]
```diff
@@ -1254,22 +1263,39 @@ public void removeOnNavigationPositionListener() {
 	@Deprecated
 	public void setNotification(int nbNotification, int itemPosition) {
 		if (itemPosition < 0 || itemPosition > items.size() - 1) {
-			Log.w(TAG, "The position is out of bounds of the items (" + items.size() + " elements)");
-			return;
+            throw new IllegalArgumentException(String.format(Locale.US, EXCEPTION_ILLEGAL_ARGUMENT, itemPosition, items.size()));
```
Comment (al-dr) [inline ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java]: IndexOutOfBound?

Comment (repitch) [inline ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/AHBottomNavigation.java]: fixed in 8f23b11

File: ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/notification/AHNotification.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            AHNotification.this.textColor = textColor;
+            return this;
+        }
+
+        public Builder setBackgroundColor(@ColorInt int backgroundColor) {
+            AHNotification.this.backgroundColor = backgroundColor;
+            return this;
+        }
+
+        public AHNotification build() {
```
Comment (deM1d) [inline ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/notification/AHNotification.java]: Builder alsway has reference to AHNotification instance. After call build() method you can change AHNotification state using builder methods. Use builder as static class and create new AHNotification instance in build() method.

File: ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/notification/AHNotificationHelper.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * @author repitch
+ */
+public final class AHNotificationHelper {
+
+    private AHNotificationHelper() {
+        // empty
+    }
+
+    public static int getTextColor(@Nullable AHNotification notification, @ColorInt int defaultTextColor) {
+        if (notification == null) {
```
Comment (deM1d) [inline ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/notification/AHNotificationHelper.java]: notification should be not null (@NonNull). IllegalStateException if it is.

File: ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/notification/AHNotification.java (hunk 5 above)
Comment (repitch) [inline ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/notification/AHNotification.java]: you're right
fixed in f9acf56

File: ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/notification/AHNotification.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+    public int getBackgroundColor() {
+        return backgroundColor;
+    }
+
+    public static AHNotification justText(String text) {
+        return newBuilder().setText(text).build();
+    }
+
+    public static Builder newBuilder() {
```
Comment (repitch) [inline ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/notification/AHNotification.java]: fixed in f9acf56
I removed this method, use `new AHNotification.Builder()` instead

File: ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/notification/AHNotificationHelper.java (hunk 6 above)
Comment (repitch) [inline ahbottomnavigation/src/main/java/com/aurelhubert/ahbottomnavigation/notification/AHNotificationHelper.java]: fixed in f9acf56

**Comment by aurelhubert** [comment]
Thanks guys, nice work, I will add the PR in the version 1.5.0

**Comment by repitch** [comment]
@aurelhubert  Thanx for the merge! Looking forward to the release.

**Comment by aurelhubert** [comment]
Released in 1.5.0.

**Pull request merged**